### ISSUE TITLE
rework wesnoth.game_config lua api

### ISSUE THREAD

**Issue by gfgtdf** [wesnoth/wesnoth#4907]
The current wesnoth game_config api is some mix of various game settings that don't really belong together, furthermore  it has some serious drawbacks in particular some important information like active_mods and difficulty is unavailable in sp.

My current suggestion (subject to change) would be to restructure it by adding wesnoth.current_scenario or just wesnoth.scenario and move thee current_scenario-related settings therr and also adding stuff like difficulty and active_mods there. 

- wesnoth
  - (current_) scenario
    (attibutes tht also go under [scenario])
    - turn_at
    - turns
    - victory_when_enemies_defeaed
    - defeat_music
    - victory_music
    - name
    - id
    - do_healing
    - theme ?
    (i think these also fit here)
    - active_modifications
    - active_ressources
    - active_era
    - active_campaign
    - difficulty
  - game_config
    (things that are defined via [game_config])
    - base_income
    - village_income
    - village_support
    - poison_amount
    - rest_heal_amount
    - recall_cost
    - kill_experience
    - combat_experience
  - mp_settings

**Comment by CelticMinstrel** [comment]
Some of the items you've listed under `game_config` are also scenario-specific and/or side-specific, like `village_income`…

The general idea of splitting this up seems fine, though. Through `wesnoth.[current_]scenario` we might be able to access the scenario's WML as well?

I'm not sure that `theme` fits under the scenario, however.

**Comment by gfgtdf** [comment]
> Some of the items you've listed under game_config are also scenario-specific and/or side-specific, like village_income…

I don't think so actually, there is village_income in [game_config] and in [side] but none in scenario, in the particular case or game_config.village_income it actually has no effect other than being a default value for the sides value so when changing it via lua there should have no effect at all, no idea why it exists, we should probably just remove village_income there.

For base_income it is afaik added to the side specific value.

> I'm not sure that theme fits under the scenario, however.

i completely agree i just put i there because its also `theme=` in [scenario], where would you put it?


> The general idea of splitting this up seems fine, though. Through wesnoth.[current_]scenario we might be able to access the scenario's WML as well?

what were you thinking of? A function that generated the wml just like it does for savefiles? I think we _could_ do that but i currently don't see a usecase for that.

**Comment by CelticMinstrel** [comment]
Ah, they were side paramaters, then, okay. Somehow I thought they also worked in `[scenario]`? Maybe because some scenarios were erroneously using them there before we got the schema validator working, or something.

> in the particular case or game_config.village_income it actually has no effect other than being a default value for the sides value so when changing it via lua there should have no effect at all, no idea why it exists, we should probably just remove village_income there.

No, don't remove it. It doesn't need to be writable, but it could be useful to be able to easily look up the default village income.

> i completely agree i just put i there because its also `theme=` in [scenario], where would you put it?

Probably in `wesnoth.interface`.

> what were you thinking of? A function that generated the wml just like it does for savefiles? I think we _could_ do that but i currently don't see a usecase for that.

Hmm, I'm not quite sure _what_ I was thinking, really. Dumping an entire save file just to look up some data clearly is a bad idea tho, so let's _not_ do that.

I guess the important thing would be the ability to look up everything that's defined in the `[scenario]` tag, except for events:

* Anything that's a key in `[scenario]` and doesn't just specify initial state.
* The music playlist (this is already available).
* The map (not as a raw string or filename, but rather as a terrain map object).
* Labels and items.
* The time schedule and any time areas, including the current time index (`current_time`, which would be the _current_ time, not the starting time).
* The list of sides (which is already available)

Note that the following probably aren't needed:
* Anything that only specifies initial state: turn_at, random_start_time, do_healing
* The story screen - it's already run its course by the time any Lua runs anyway, right?
* Map generator data (the map is already generated by the time any Lua runs)
* Events

The API needs to present the _current_ state of the scenario. It doesn't _hurt_ to have members specifying what the initial state was, but it's not as useful. (I'd also give them different names if you want to include them – `start_time` instead of `current_time`, though `current_time` would still exist; `start_turn` instead of `turn_at`.)

**Comment by gfgtdf** [comment]
The thing is, technicially a lot of things, nearly everything, would  fit into 'current_scenario': units, map, events, labels, music etc. since they describe the current state of the scenario. But my plan here really wasn't to move all of these. It's just that currently there are  a few keys in game_config that don't really belong there in my opinion.

So my plan is still to just move things there that doesn't have a place already, which probably leaves the things i mentioned in my first post. 

> if you want to include them – start_time instead of current_time, though current_time would still exist; start_turn instead of turn_at.)

I never said i wanted to include attributes that specifiy the initial state of the scenario. And i see little reason to call it `current_turn` instead of  `turn_at` if `[scenario]` uses `turn_at` already (i don't have a really strong opinion on this one but unless some other api already uses `current_turn` i'd probably go for `turn_at`)

**Comment by CelticMinstrel** [comment]
Does 97a96cbbd15ccf76c1444f81b444f82ac2ecd724 cover this or is there still more to do?

**Comment by gfgtdf** [comment]
Nothing that i can think of right now (i'm still thinking there is _somethign_ else i wanted to move out of `[mp_settings]` but i cannot think of what it is right now)
